test/distributed/_tensor/test_dtensor.py

```diff
@@ -704,5 +707,265 @@ def test_split_tensor(self) -> None:
+    view_1 = torch.ops.aten.view.default(_to_copy, [2, 4]);  _to_copy = None
+    view_2 = torch.ops.aten.view.default(primals_2, [4, 2]);  primals_2 = None
+    _to_copy_1 = torch.ops.aten._to_copy.default(view_2, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  view_2 = None
+    view_3 = torch.ops.aten.view.default(_to_copy_1, [4, 2]);  _to_copy_1 = None
+    all_gather_into_tensor = torch.ops.c10d_functional.all_gather_into_tensor.default(view_3, 'ptd:0', [0, 1], 2)
+    wait_tensor = torch.ops.c10d_functional.wait_tensor.default(all_gather_into_tensor);  all_gather_into_tensor = None
+    split = torch.ops.aten.split.Tensor(wait_tensor, 4);  wait_tensor = None
+    getitem = split[0]
+    getitem_1 = split[1];  split = None
+    cat = torch.ops.aten.cat.default([getitem, getitem_1], 1);  getitem = getitem_1 = None
+    mm_1 = torch.ops.aten.mm.default(view_1, cat);  cat = None
+    view_4 = torch.ops.aten.view.default(mm_1, [2, 4]);  mm_1 = None
+    t_1 = torch.ops.aten.t.default(view_1);  view_1 = None
+    t_3 = torch.ops.aten.t.default(view_3);  view_3 = None
+    clone = torch.ops.aten.clone.default(t_3, memory_format = torch.contiguous_format);  t_3 = None
+    return [view_4, t_1, clone]""",  # noqa: B950
+        )
+
+    # Same test as above, but testing the backwards.
+    def test_compile_dtensor_redistribute_backward(self):
+        mesh = DeviceMesh(self.device_type, torch.arange(self.world_size))
+
+        def fn(x, y):
+            dt = DTensor.from_local(x.reshape(2, 4), mesh, [Shard(0)], run_check=False)
+            dt2 = DTensor.from_local(y.reshape(4, 2), mesh, [Shard(1)], run_check=False)
+            dt_out = torch.matmul(dt, dt2)
+            dt_out_redistribute = dt_out.redistribute(mesh, [Replicate()])
+            return dt_out.to_local()
+
+        opt_fn = torch.compile(fn, backend=aot_eager_graph, fullgraph=True)
+
+        x_ref = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        y_ref = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        ref = fn(x_ref, y_ref)
+
+        x = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        y = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        res = opt_fn(x, y)
+
+        self.assertEqual(res, ref)
+
+        # Now run and assert the backward + gradients
+        ref.sum().backward()
+        res.sum().backward()
+
+        # The bw graph here is pretty messy (there are unnecessary cat and split calls?)
+        # But the important ting to note is:
+        # (1) we have a c10d_functional op in the graph, thanks to DTensor
+        # (2) we have a wait() call in the graph that inductor can reorder
+        # (3) We have a mm() in the graph (from the user code)
+        self.assertExpectedInline(
+            bw_graph_cell[0].code.strip(),
+            """\
+def forward(self, t_1, clone, tangents_1):
```

Comment:
do we need to assert the graph here? a bit worried that the graph assertion might be hard to navigate when the pt2 stack have new changes breaks the assertion (i.e. adding new op decomp), I am wondering if we should just do `fullgraph=True` and assert the output?

Reply:
hmm, the nice thing about the graph assert IMO is that someone can at-a-glance look at the tests and understand what the workflow of compiling DTensors is ("this is what matmul + redistribute in the backward desugars into").

If we update a decomp somewhere, updating the tests is pretty easy - running `EXPECTTEST_ACCEPT=1 python test/distributed/_tensor/test_dtensor.py` will automatically update the file.

But if you'd still prefer to keep these out of the test file, lmk and I'm happy to r3emove.

Reply:
Oh I see, that's good then, let's keep it! Might worth adding this `EXPECTTEST_ACCEPT=1 python test/distributed/_tensor/test_dtensor.py` command in the test comment so ppl know how to fix those when they saw failures :)